### ISSUE TITLE
Simple TFLite UNet slower on mobile GPU than CPU

### ISSUE THREAD

**Issue by jakobknapp** [tensorflow/tensorflow#41769]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution: Linux Ubuntu 18.04, Android 8.0.0
- Mobile device if the issue happens on mobile device: Huawei P20 Lite (ANE-LX1)
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b
- Python version: 3.6.9
- CUDA/cuDNN version: CUDA 10.1
- GPU model: Mali-T830 MP2

**Describe the current behavior**

I have written a simple image segmentation model based on UNet in Keras, which I want to use on an Android device (Huawei P20 Lite). It consists solely of Conv2D, Conv2DTranspose, and Concatenate layers, which should all be supported on mobile GPUs by TFLite according to the [documentation](https://www.tensorflow.org/lite/performance/gpu_advanced#supported_ops). The model has ~1.9M parameters and expects input tensors of shape [1, 224, 224, 3] and type float32. It is not trained yet, as I want to check its performance first before addressing the accuracy. I exported the model using

```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()
with tf.io.gfile.GFile('trained models/keras_unet.tflite', 'wb') as f:
  f.write(tflite_model)
```
You can find the exported TFLite model attached further down.

On the CPU of my device, inference takes about 1.3s on average using 6 threads. However, when employing the GPU delegate using

```
tfliteModel = FileUtil.loadMappedFile(this, MODEL_NAME_TF);
GpuDelegate gpu_del = new GpuDelegate();
tfliteOptions.addDelegate(gpu_del);
tflite = new Interpreter(tfliteModel, tfliteOptions);
```

the performance drops to about 3.5s per inference, which is more than 2.5 times slower compared to CPU. 

**Describe the expected behavior**
As the UNet model only consists of supported operations, the mobile GPU supports OpenGL ES 3.2, and model export and import as well as GPU delegate creation have been done as suggested in the [TFLite Android Tutorial](https://codelabs.developers.google.com/codelabs/recognize-flowers-with-tensorflow-on-android/#0) and the [GPU delegate Guide](https://www.tensorflow.org/lite/performance/gpu#trying_the_gpu_delegate_on_your_own_model), I would expect the UNet to infer faster on the GPU than on the CPU.

I would be glad if you could point out any implementation issues on my side that could potentially cause the comparably worse performance on the GPU. The guides and tutorials unfortunately do not offer any deeper insight. Is it possible that the model is simply too large for my GPU (and would there be an error if this was the case)?

**Standalone code to reproduce the issue**
The following model exhibits the described behavior: [tflite_model.zip](https://github.com/tensorflow/tensorflow/files/4982040/tflite_model.zip)

**Other info / logs**
I have previously implemented the exact same model in PyTorch, the performance on the CPU is about the same. I then switched to Keras to utilize TFLite's GPU delegate. I used a subclassed approach to develop the UNet-like model in Keras.

I have read [here](https://www.tensorflow.org/lite/performance/gpu#tips_for_optimization) that it might be beneficial to use a Tensor with c=4 instead to avoid memory copies. I tried adding a fourth channel to the model and the input images (shape [1, 224, 224, 4]), which made performance even worse.

The performance has been tested using a custom Android application. The timings mentioned above refer to the following call:
`tflite.run(inputTensorBuffer.getBuffer(), outputProbabilityBuffer.getBuffer().rewind());`
I also attempted to profile the TFLite model using the [TFLite Model Benchmark Tool](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/benchmark), but the Tool failed to build correctly using Bazel. If this is of relevance for solving the problem at hand, I will file another issue for that.

Finally, here are the outputs produced by the Python script for exporting the model and the Android test application:
[python_out.txt](https://github.com/tensorflow/tensorflow/files/4982006/python_out.txt)
[android_out.txt](https://github.com/tensorflow/tensorflow/files/4982010/android_out.txt)

**Comment by renjie-liu** [comment]
Hi Chao, can you help take a look?

thanks

**Comment by multiverse-tf** [comment]
Hi Jakob, could you download the pre-built model benchmark tool from [here](https://storage.googleapis.com/tensorflow-nightly-public/prod/tensorflow/release/mobile/android_lite/nightly/806/20200726-221609/benchmark_tflite_model), then run it on your device and share detailed profiling info here?

On my side, I tried this benchmark tool with this model (keras_unet.tflite) on a Pixel 3 and found that gpu gives much better perf. than the cpu one. If using the OpenGL backend, it's about ~150ms. If using the OpenCL backend, it's about ~32ms.

**Comment by jakobknapp** [comment]
Hi Chao, thanks for looking into that issue!

Your results seem to suggest that the model should be alright and the bad performance is either due to some error I have in my Android application or because my phone can't keep up.

I used the pre-built benchmark tool you linked and the results are similar to my Android application. While inference is a bit faster in total, GPU inference is still slower by the same factor relative to the CPU.

Average Inference time on **CPU**: ~1s

```
WARNING: linker: "/data/data/org.picsart.bgremoval/tf_test/benchmark_tflite_model" unused DT entry: type 0xf arg 0x930
STARTING!
Duplicate flags: num_threads
Log parameter values verbosely: [0]
Num threads: [6]
Graph: [keras_unet.tflite]
Enable op profiling: [1]
CSV File to export profiling data to: [benchmark_cpu.csv]
#threads used for CPU inference: [6]
Loaded model keras_unet.tflite
INFO: Initialized TensorFlow Lite runtime.
The input model file size (MB): 7.70576
Initialized session in 3.487ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=1 curr=1289876

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=50 first=1064709 curr=1013050 min=997794 max=1064709 avg=1.01013e+06 std=10693

Inference timings in us: Init: 3487, First inference: 1289876, Warmup (avg): 1.28988e+06, Inference (avg): 1.01013e+06
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Peak memory footprint (MB): init=2.11719 overall=314.234
```

Average Inference time on **GPU** using OpenGL backend: ~2.4s

```
WARNING: linker: "/data/data/org.picsart.bgremoval/tf_test/benchmark_tflite_model" unused DT entry: type 0xf arg 0x930
STARTING!
Duplicate flags: num_threads
Log parameter values verbosely: [0]
Graph: [keras_unet.tflite]
Enable op profiling: [1]
CSV File to export profiling data to: [benchmark_gpu.csv]
Use gpu: [1]
Loaded model keras_unet.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite delegate for GPU.
ERROR: TensorDescriptor don't have selector with name - Slice
ERROR: Falling back to OpenGL
INFO: Initialized OpenGL-based API.
INFO: Created 1 GPU delegate kernels.
Applied GPU delegate, and the model graph will be completely executed by the delegate.
The input model file size (MB): 7.70576
Initialized session in 2101.73ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=1 curr=2503969

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=50 first=2345432 curr=2325535 min=2284622 max=2461721 avg=2.35429e+06 std=31681

Inference timings in us: Init: 2101733, First inference: 2503969, Warmup (avg): 2.50397e+06, Inference (avg): 2.35429e+06
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Peak memory footprint (MB): init=102.875 overall=102.875
```

However, I could not get it to run using the OpenCL GPU backend you mentioned. I have read that this should be the default behavior, but if no backend is specified, it falls back to the OpenGL backend as seen above.
I have also tried enforcing the OpenCL backend by using `--gpu_backend="cl"`, which failed:

```
WARNING: linker: "/data/data/org.picsart.bgremoval/tf_test/benchmark_tflite_model" unused DT entry: type 0xf arg 0x930
STARTING!
Duplicate flags: num_threads
Log parameter values verbosely: [0]
Graph: [keras_unet.tflite]
Use gpu: [1]
GPU backend: [cl]
Loaded model keras_unet.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite delegate for GPU.
ERROR: TfLiteGpuDelegate Init: TensorDescriptor don't have selector with name - Slice
INFO: Created 0 GPU delegate kernels.
ERROR: TfLiteGpuDelegate Prepare: delegate is not initialized
ERROR: Node number 24 (TfLiteGpuDelegateV2) failed to prepare.

ERROR: Restored original execution plan after delegate application failure.
Failed to apply GPU delegate.
Benchmarking failed.
```

Here is the OP Profiling Output in case it provides any additional insight:  [benchmark_op_profiling.zip](https://github.com/tensorflow/tensorflow/files/4989590/benchmark_op_profiling.zip)

So it looks like both the Model and the Android application are fine, and the issue is somehow related to OpenCL on my device, as it works properly on yours.

**Comment by multiverse-tf** [comment]
Hi Jakob,

Based on outputs of the benchmark tool you pasted above (i.e. "ERROR: TfLiteGpuDelegate Init: TensorDescriptor don't have selector with name - Slice"), I agree with you that the OpenCL lib on your device is the culprit to the failed OpenCL GPU delegate run.  Could you tell the OpenCL version on your device?

It's also not clear to me why the OpenGL GPU performance is worse than that of CPU (which used 6 threads based on the output). It might be related to the OpenGL library on your device as well. Just wondering do you have another different Android phone to check the performance? 

Added @impjdi to shed additional lights on this issue.

**Comment by jakobknapp** [comment]
Hi,

yes, I could get my hands on a Samsung Galaxy S6 for comparison. Unfortunately, I do not have root access and also cannot use run-as, so I could not use your benchmark tool on the S6. However, my Android test-app behaves as expected:
~710ms/frame on GPU vs. ~850ms/frame on CPU

So the S6 seems to be faster than the P20 Lite in general and inference is faster on the GPU than on the CPU. On the other hand, it also seems to use OpenGL instead of OpenCL according to the Logs:

```
I/tflite: Created TensorFlow Lite delegate for GPU.
I/tflite: Initialized TensorFlow Lite runtime.
D/libEGL: loaded /vendor/lib64/egl/libGLES_mali.so
E/libEGL: validate_display:99 error 3008 (EGL_BAD_DISPLAY)
E/libEGL: call to OpenGL ES API with no current context (logged once per thread)
I/tflite: Initialized OpenGL-based API.
I/tflite: Created 1 GPU delegate kernels.
```

The OpenGL ES/OpenCL versions of the two phones (Huawei P20 Lite with Mali-T830 GPU vs. Samsung Galaxy S6 with Mali-T760 GPU) are almost identical. I used [this app](https://play.google.com/store/apps/details?id=nl.svdree.glesinfo&hl=en) and [this app](https://play.google.com/store/apps/details?id=com.robertwgh.opencl_z_android&hl=en) for checking:

### **P20 Lite:**

![P20Lite Drivers](https://user-images.githubusercontent.com/44022022/88913407-1ee1da00-d261-11ea-8831-24e7113a28fd.png)

### **Galaxy S6:**

![S6 Drivers](https://user-images.githubusercontent.com/44022022/88913425-26a17e80-d261-11ea-8260-950b8ea61b75.png)

The main difference is that the S6 has 8 compute units while the P20 Lite has only 2. According to the Performance tab, the S6 (right) also seems to have a higher float vec4 throughput:

![Performance Comparison](https://user-images.githubusercontent.com/44022022/88917154-9155b880-d267-11ea-900a-9b966932720e.jpg)

**Comment by multiverse-tf** [comment]
Jakob, many thanks for pasting raw perf. specs of these two phones! As you said, the GPU on S6 has a higher float vec4 throughput (>6X over the P20 Lite). I also noted that the host<->device memory BW on S6 is ~2X higher as well. Therefore, I think it's expected that the TFLite GPU performs much better on S6 than that on P20 Lite. It might be safe to conclude that it's not a good option to use GPU delegate on P20 Lite generally.

**Comment by jakobknapp** [comment]
Hi Chao, thanks for checking! Yes, that seems reasonable. Guess I have to stick with the CPU for now.

One last question before I close this issue: Which specs of the mobile GPU do I have to pay special attention to when considering to use a GPU delegate? Does TFLite extensively utilize vec4 computations in particular?

**Comment by multiverse-tf** [comment]
> Hi Chao, thanks for checking! Yes, that seems reasonable. Guess I have to stick with the CPU for now.
> 
> One last question before I close this issue: Which specs of the mobile GPU do I have to pay special attention to when considering to use a GPU delegate? 
This is a great question! And we are working to resolve this issue w.r.t. whether to recommend using GPU delegate. At the moment, I really don't have a good answer to this. But I think GPU delegate generally delivers better performance on most recent mid-range+ phones with Android 9+. 

Does TFLite extensively utilize vec4 computations in particular?
I don't know. @impjdi, could you help w/ this question? Thx!

**Comment by impjdi** [comment]
Which spec is difficult to answer.  The performance depends on many factors such as what kind of GPU you have (typically Adreno vs Mali), how large your tensors are (both runtime and trained weights), how fast your ops are (if you have too many small ops, performance hurts), and how many ops you have etc.  You can take a quick look at the kernel implementations under `gpu/delegates/{gl,cl}/kernels`, but yes, we use `vec4` a lot.
